Src/FluentAssertions/Formatting/FormattedObjectGraph.PossibleMultilineFragment.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+
+namespace FluentAssertions.Formatting;
+
+/// <summary>
+/// Implementation of the contained PossibleMultilineFragment.
+/// </summary>
+public partial class FormattedObjectGraph
+{
+    /// <summary>
+    /// Write fragments that may be on a single line or span multiple lines,
+    /// and this is not known until later parts of the fragment are written.
+    /// </summary>
+    internal record PossibleMultilineFragment
+    {
+        private readonly FormattedObjectGraph formattedGraph;
+        private readonly int startingLineBuilderIndex;
+        private readonly int startingLineCount;
+
+        public PossibleMultilineFragment(FormattedObjectGraph formattedGraph)
+        {
+            this.formattedGraph = formattedGraph;
+            startingLineBuilderIndex = formattedGraph.lineBuilder.Length;
+            startingLineCount = formattedGraph.lines.Count;
+        }
+
+        /// <summary>
+        /// Write the fragment at the position the graph was in when this instance was created.
+        ///
+        /// <para>
+        /// If more lines have been added since this instance was created then write the
+        /// fragment on a new line, otherwise write it on the same line.
+        /// </para>
+        /// </summary>
+        internal void AddFragmentAtStart(string fragment)
+        {
+            if (FormatOnSingleLine)
```

Comment:
Hmm.  My thinking is that if one of the children will be formatted over multiple lines then they all will.  Is that valid, or are there cases where one enumerable item will be on a single line but others (of the same type) will be on multiple lines?

The idea is to format one child, and then decide if the whole thing will be single-line or multi-line based on that.

Reply:
Argh, pathological edge case! :) 

You are right, if the first item in a list has an overidden short `ToString` and a later item is formatted with its members then the algorithm didn't quite work.  Fixing that uncovered a formatting bug with nested list/objects as well - *and* I think I found an existing quirk some enumerable assertions.

I have now [added a test](https://github.com/fluentassertions/fluentassertions/blob/9ad585b6f0c735c23f98f03fe5203de94bee4872/Tests/FluentAssertions.Specs/Primitives/ReferenceTypeAssertionsSpecs.cs#L57) for a mixture of overriden and not-overriden `ToString` classes:

    Expected subject to be empty, but found 
    {
        Simple(Hello), 
        FluentAssertions.Specs.Primitives.Complex
        {
            Statement = "goodbye"
        }
    }.

I have also [added a test](https://github.com/fluentassertions/fluentassertions/blob/9ad585b6f0c735c23f98f03fe5203de94bee4872/Tests/FluentAssertions.Specs/Formatting/FormatterSpecs.cs#L318) that collections with complex objects that contain collections of complex objects format properly:

    Expected stuff to be a collection with 1 item(s), but 
    {
        {
            Description = "absent"
        }, 
        {
            Description = "absent"
        }
    }
    contains 1 item(s) more than

    {
        {
            ComplexChildren = 
            {
                {
                    Property = "hello"
                }, 
                {
                    Property = "goodbye"
                }
            }
        }
    }.

The required a change to the enumerable formatting algorithm, so please have another look at [`EnumerableValueFormatter`](https://github.com/fluentassertions/fluentassertions/blob/9ad585b6f0c735c23f98f03fe5203de94bee4872/Src/FluentAssertions/Formatting/EnumerableValueFormatter.cs#L30) and [`PossibleMultilineFragment`](https://github.com/fluentassertions/fluentassertions/blob/9ad585b6f0c735c23f98f03fe5203de94bee4872/Src/FluentAssertions/Formatting/FormattedObjectGraph.cs#L173).

---

Whilst digging into all of this this I noticed this:

    var expected = new[] { 1, 2, 3, };
    var actual = new[] { 1, 2, 3, 4 };
    actual.Should().BeEquivalentTo(expected);

    Expected actual to be a collection with 3 item(s), but {1, 2, 3, 4}"
    "contains 1 item(s) more than"
    "{1, 2, 3}.

Those strange quotation marks are because `AssertCollectionHasNotTooManyItems` was adding `Environment.NewLine` (which is a string) as an argument to the formatter, and `StringValueFormatter` wraps all strings in quotes.  I [have changed it](https://github.com/fluentassertions/fluentassertions/pull/2144/commits/9ad585b6f0c735c23f98f03fe5203de94bee4872) so that the NewLines are part of the message template instead.